lib/src/connector/ischool_plus_connector.dart

```diff
@@ -58,13 +66,27 @@ class ISchoolPlusConnector {
         final value = node.attributes['value'];
         data[name] = value;
       }
+
+      // Step 1
       final jumpUrl = tagNode.getElementsByTagName("form")[0].attributes["action"];
-      parameter = ConnectorParameter(jumpUrl);
+      parameter = ConnectorParameter("${NTUTConnector.host}$jumpUrl");
       parameter.data = data;
 
       Response<dynamic> jumpResult = (await Connector.getDataByPostResponse(parameter));
+      tagNode = html.parse(jumpResult.data.toString().trim());
+      nodes = tagNode.getElementsByTagName("a");
+      final redirectUrl = nodes[0].attributes["href"];
```

Comment:
```suggestion
      final redirectUrl = nodes.first.attributes["href"];
      if (redirectUrl == null) {
        return ISchoolPlusConnectorStatus.loginFail;
      }
```

Suggest adding null check to this section, even we haven't enabled the null safety feature on this file.
The `Node.attributes` is a `LinkedHashMap<Object, String>`, which inherited the `[]` from the `Map` class like this:
```dart
  /// The value for the given [key], or `null` if [key] is not in the map.
  ///
  /// Some maps allow `null` as a value.
  /// For those maps, a lookup using this operator cannot distinguish between a
  /// key not being in the map, and the key being there with a `null` value.
  /// Methods like [containsKey] or [putIfAbsent] can be used if the distinction
  /// is important.
  V? operator [](Object? key);
```

In addition, suggest using `List.first` to replace with `[0]`, since there error type will be different when there's no element inside the list.

- `[0]`: `RangeError (index): Index out of range`
- `List.first`: `Uncaught Error: Bad state: No element` (see `IterableElementError.noElement`)

Seems the second error is more preferable, because it tells us "No element" rather than just saying "index out of range".

Reply:
Fix on 8a80309

Reply:
```dart
      if (redirectUrl == null) {
        // maybe some log code here...
        return ISchoolPlusConnectorStatus.loginFail;
      }
```

[Question]
Is there any reason that we skip the null check here ?

Reply:
At first, I think it should exist or the step should crash. But I don't have any idea how not to change too much of the code and check whether the step is good or died.

Maybe it will optimize the "step flow" with linked-structure and provide the failed situation if it dies.

```dart
RunStep("step 1", function)
    .ThenRunStep("step 2", function)
    .ThenRunStep("step 3", function)
// ...
```

Reply:
The reason why I consider handling the null case here is because:

Catching the problem earlier may help us find the error more accurate.
If we expect it throws an error or exception to the try...catch, sometimes we may not easy to distinguish where this error comes from. (avoid using try...catch or minimize the try...catch scope)

Howerver, this file was created at a very early time, which may need to be refactored in the future.
On top of this reason, you may consider using existing strategy & structure to prevent breaking changes here.

Reply:
As you mentioned at https://github.com/NEO-TAT/tat_flutter/pull/206#discussion_r1367738948, 
I will recommend seeing how flutter official team do on similar case.
For example, the `setState` operation: https://github.com/flutter/flutter/blob/025550f26ac9fe31ba2c7d649a33255cbccff188/packages/flutter/lib/src/widgets/framework.dart#L1164

Both of our case and `setState` include multiple different step inside.
As you can see, however, they catch each error one by one, as early as possible, and seldom use try...catch structure.

In your previous provided code, if we considered using one try...catch to wrap all steps, we may need to throw different error/exception in each step of the chain.

I believe this article: https://www.sandromaglione.com/techblog/either-error-handling-functional-programming describes quite great about error handling.

Reply:
Thanks for providing the information. It will be great to improve the implementation in the future.

Reply:
Since we can't describe the step easily and make it easy to maintain, I would like to introduce the structure below.

```dart
try {
  Step("Post something to ssoIndex", () {
    // do something
  });
  Step("Post something to ...", () {
    // do something
  });
  Step("Get something to ...", () {
    // do something
    throw StepException("Fetch the redirect URL failed. since it is not login2.php in the end.");
  });
} catch (StepException e){
   // We will declare the step is failed in here and it should be explain the reason why it failed with StepException class.
}
```

Reply:
This proposal seems better when performing on single-thread (no parallel or concurrency cases), in my opinion.
If so, we may need to outline a generic Step with error structure for upcoming cases :)